packages/@aws-cdk/pipelines/lib/pipeline.ts

```diff
@@ -117,12 +117,6 @@ export class CdkPipeline extends Construct {
     if (!props.sourceAction && (!props.codePipeline || props.codePipeline.stages.length < 1)) {
       throw new Error('You must pass a \'sourceAction\' (or a \'codePipeline\' that already has a Source stage)');
     }
-    if (!props.synthAction && (!props.codePipeline || props.codePipeline.stages.length < 2)) {
```

Comment:
You are technically correct that this is not necessary, and I see you have a use case where they aren't necessary.

Your use case is outside the 99% though, so I'm hesitant to give up a guard rail that would help all people out there who aren't in your position.

How about a compromise: providing your own `pipeline` is an "I know what I'm doing" switch, so we don't require 2 stages in the existing pipeline (but if no pipeline is given, we still require the synthAction?).

Reply:
Hi @rix0rrr, thanks for your reply.

> How about a compromise: providing your own pipeline is an "I know what I'm doing" switch, so we don't require 2 stages in the existing pipeline (but if no pipeline is given, we still require the synthAction?).
I actually believe that's the behaviour this PR introduces: You must provide either `sourceAction` or `codePipeline`, otherwise the validation fails on line 117. If you provide `sourceAction`, then you also need to provide `synthAction`, otherwise the validation fails on line 110. That means the only way of not providing the `synthAction` is to provide a `codePipeline` that already has the Source stage. Does that make sense?

Reply:
If you like, I am happy to slightly reorder the validation steps to make the dependencies more obvious.